Test cases and example fixes for tf_compile_graph with recurrent step

JackTemaki opened the pull request:
Tries to fix issues when compiling with `--rec_step_by_step` by doing the following changes:

 - Adapted the code to the new internal structure of the rec layer
 - Resolved the dependency issue on the target length for the `cond` state var
 - Added two test cases with a simple search decoder, one for default compilation, one for compiling with rec-step-by-step

albertz commented on tools/compile_tf_graph.py:
```diff
@@ -461,13 +460,25 @@ def add_stochastic_var(self, name):
     """
-    :param str|dict[str] unit:
+    Replace the unit by an identical SubnetworkRecCellSingleStep instance
+    Only allows for SubnetworkRecCell, not strings as the super method does
```
We should not make a new instance/copy.
I'm pretty sure there is several code which expects that it stays the same instance.
I'm also pretty sure such copy could potentially introduce some very nasty edge cases and buggy behavior.

Rather, this `RecStepByStepLayer` should directly create a `SubnetworkRecCellSingleStep` instance in the first place, instead of a `_SubnetworkRecCell`.
This creation happens in `RecLayer.transform_config_dict` via `subnet = _SubnetworkRecCell(...)`.
I just pushed 416209a2d8929667b7e8011d0413d0706d33ae0a which allows you to easily make use of `SubnetworkRecCellSingleStep` in the same way.

albertz replied:
I think this whole function `_get_cell` can be removed here then.

albertz replied:
(So you need to rebase the branch / PR on current master.)

JackTemaki replied:
I did not want to mess with this without your comments on that, so I will check now how that works with your change.

albertz commented on tools/compile_tf_graph.py:
```diff
@@ -633,11 +644,6 @@ def cond(i, net_vars, acc_tas, seq_len_info=None)
-      rec_layer.create_state_var("cond", tf.constant(True))
-      res = cond(i, net_vars, acc_tas, seq_len_info)
-      rec_layer.set_state_var_final_value("cond", res)
-
```
You talked about this. But what exactly was the problem? Can you post some details about the error you get because of this?

Also, just removing this is not really an option anyway. We could maybe make it optional (some option like `add_cond` or so). But before we do that, we should understand what the real problem is. Because I would not expect that this should cause any problem. It should not be a problem to just have this always in the graph, such that it can be used when needed/wanted.

albertz commented on returnn/tf/layers/rec.py:
```diff
@@ -985,6 +985,8 @@ def __init__(self, net_dict, source_data, time_dim_tag, rec_layer_name, parent_n
     self.parent_net = parent_net
     self.net_dict = safe_deep_copy(net_dict)
+    self.rec_layer_name = rec_layer_name
+    self.parent_get_layer = parent_get_layer
```
I think copying `parent_get_layer` is a bad idea (can potentially lead to very nasty problems). You only need that for your copy of `_SubnetworkRecCell`? Then this can be removed after the other cleanup I suggested before.

albertz commented on tests/test_tools.py:
```diff
@@ -0,0 +1,82 @@
+
+
+
+
```
I think there are too many empty lines at the end of the file.

albertz commented:
The tools-test file should also be added to the GitHub actions.

What is the state about the tests? They already run through?

JackTemaki commented:
> The tools-test file should also be added to the GitHub actions.
> 
> What is the state about the tests? They already run through?

The tests did run through with the current code, after rebasing and addressing your comments I will add them to the GitHub actions.

JackTemaki commented:
@albertz FYI:

The `cond` code causes the following error with the test case, without it works without crash now. I did not look into this yet.
```python
EXCEPTION
Traceback (most recent call last):
  File "/home/nick/github/returnn/tools/compile_tf_graph.py", line 750, in <module>
    line: main(sys.argv)
    locals:
      main = <local> <function main at 0x7f417d1b1310>
      sys = <local> <module 'sys' (built-in)>
      sys.argv = <local> ['tools/compile_tf_graph.py', '--output_file', '/tmp/tmpf6kfnzbg/graph.metatxt', '--rec_step_by_step', 'output', '/tmp/tmpf6kfnzbg/returnn.config'], len = 6, _[0]: {len = 25}
  File "/home/nick/github/returnn/tools/compile_tf_graph.py", line 679, in main
    line: RecStepByStepLayer.post_compile(
            rec_layer_name=args.rec_step_by_step, network=network, output_file_name=args.rec_step_by_step_output_file)
    locals:
      RecStepByStepLayer = <global> <class '__main__.RecStepByStepLayer'>
      RecStepByStepLayer.post_compile = <global> <bound method RecStepByStepLayer.post_compile of <class '__main__.RecStepByStepLayer'>>
      rec_layer_name = <not found>
      args = <local> Namespace(config='/tmp/tmpf6kfnzbg/returnn.config', train=0, eval=0, search=0, verbosity=4, summaries_tensor_name=None, rec_step_by_step='output', rec_step_by_step_output_file=None, output_file='/tmp/tmpf6kfnzbg/graph.metatxt', output_file_model_params_list=None, output_file_state_vars_list=None)
      args.rec_step_by_step = <local> 'output', len = 6
      network = <local> <TFNetwork 'root' train=False>
      output_file_name = <not found>
      args.rec_step_by_step_output_file = <local> None
  File "/home/nick/github/returnn/tools/compile_tf_graph.py", line 277, in RecStepByStepLayer.post_compile
    line: next_step_ops.append(var.final_op())
    locals:
      next_step_ops = <local> []
      next_step_ops.append = <local> <built-in method append of list object at 0x7f4266f7bcc0>
      var = <local> <StateVar 'cond', shape Data(name='cond', dtype='bool', batch_shape_meta=[]), initial <tf.Tensor 'output/rec/cond/Const:0' shape=() dtype=bool>>
      var.final_op = <local> <bound method RecStepByStepLayer.StateVar.final_op of <StateVar 'cond', shape Data(name='cond', dtype='bool', batch_shape_meta=[]), initial <tf.Tensor 'output/rec/cond/Const:0' shape=() dtype=bool>>>
  File "/home/nick/github/returnn/tools/compile_tf_graph.py", line 396, in RecStepByStepLayer.StateVar.final_op
    line: assert not path, "There should be no path from extern data to this final op value, but there is: %r" % (path,)
    locals:
      path = <local> [<tf.Operation 'output/rec/subnet_base/classes_transpose' type=Transpose>, <tf.Operation 'output/rec/subnet_base/check_data_len/Shape' type=Shape>, <tf.Operation 'output/rec/subnet_base/check_data_len/strided_slice' type=StridedSlice>, <tf.Operation 'output/rec/cond/loop_cond/i_less_max_seq_len' ...
AssertionError: There should be no path from extern data to this final op value, but there is: [<tf.Operation 'output/rec/subnet_base/classes_transpose' type=Transpose>, <tf.Operation 'output/rec/subnet_base/check_data_len/Shape' type=Shape>, <tf.Operation 'output/rec/subnet_base/check_data_len/strided_slice' type=StridedSlice>, <tf.Operation 'output/rec/cond/loop_cond/i_less_max_seq_len' type=Less>, <tf.Operation 'output/rec/cond/loop_cond/LogicalAnd' type=LogicalAnd>]
```

albertz commented:
You write this in the description of the PR:

> Tries to fix issues

So is this still incomplete? Or is the description wrong/outdated? Please describe in the description what the PR is supposed to do (when it is complete).

In comments, you can then write that it is still incomplete, or what the current problems are.

You can later also update the description of the PR, or mark comments as resolved or outdated. That makes it much easier to understand the current state.

> The `cond` code causes the following error ...

This is resolved now? Or still a problem? If it is still a problem, why is the PR marked as ready?

> tools-test file should also be added to the GitHub actions

This is not yet done? So it is not ready yet?

JackTemaki commented:
> You write this in the description of the PR:
> 
> > Tries to fix issues
> 
> So is this still incomplete? Or is the description wrong/outdated? Please describe in the description what the PR is supposed to do (when it is complete).

I would still call it "try", because this only covers one specific test cases. So maybe it works now for everyone, maybe not.

I will update the description to summarize what I did.

> 
> In comments, you can then write that it is still incomplete, or what the current problems are.
> 
> You can later also update the description of the PR, or mark comments as resolved or outdated. That makes it much easier to understand the current state.
> 
> > The `cond` code causes the following error ...
> 
> This is resolved now? Or still a problem? If it is still a problem, why is the PR marked as ready?

Yes, this is resolved.

> 
> > tools-test file should also be added to the GitHub actions
> 
> This is not yet done? So it is not ready yet?

You are right, I forgot to add the github action.
